### PR TITLE
Add support for HTE501

### PR DESCRIPTION
**Pull request by Stock-M**
# What does this implement/fix?

Add support for HTE501 sensor.

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Other

**Related issue or feature (if applicable):** fixes 

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** esphome/esphome-docs#2241

## Test Environment

- [x] ESP32
- [ ] ESP32 IDF
- [ ] ESP8266

## Example entry for `config.yaml`:

```yaml

sensor:
  - platform: hte501
    temperature:
      name: "Office Temperature"
    humidity:
      name: "Office Humidity"
    address: 0x40
    update_interval: 60s

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

### CODE REVIEW

File: tests/test1.yaml [hunk 1]
```diff
@@ -628,6 +628,13 @@ sensor:
     temperature:
       name: "Honeywell temperature"
     cs_pin: GPIO5
+  - platform: hte501
+    temperature:
+      name: "Office Temperature 2 "
+    humidity:
+      name: "Office Humidity 1"
+    address: 0x40
+    i2c_id: i2c_bus
```
Comment (jesserockz) [inline tests/test1.yaml]:
```suggestion
      name: Office Temperature 2
    humidity:
      name: Office Humidity 1
    address: 0x40
    i2c_id: i2c_bus
```

File: esphome/components/hte501/hte501.cpp [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+      {
+        crc_val = ((crc_val << 1) ^ 0x31);
+      } else {
+        crc_val = (crc_val << 1);
+      }
+      cur_val = cur_val << 1;
+    }
+  }
+  return crc_val;
+}
```
Comment (jesserockz) [inline esphome/components/hte501/hte501.cpp:88]: You should use `crc8` that is already in helpers and just pass the part of the buffer you want to calculate on with the length.

File: esphome/components/hte501/hte501.cpp [hunk 3]
```diff
@@ -0,0 +1,90 @@
+#include "hte501.h"
+#include "esphome/core/log.h"
```
Comment (jesserockz) [inline esphome/components/hte501/hte501.cpp:2]:
```suggestion
#include "esphome/core/log.h"
#include "esphome/core/helpers.h"
```

File: esphome/components/hte501/hte501.cpp [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+  this->write(address_1, 2, true);
+  this->set_timeout(50, [this]() {
+    uint8_t i2c_response[6];
+    this->read(i2c_response, 6);
+    if (i2c_response[2] != calc_crc8_(i2c_response, 0, 1) && i2c_response[5] != calc_crc8_(i2c_response, 3, 4)) {
+      this->error_code_ = CRC_CHECK_FAILED;
+      this->status_set_warning();
+      return;
+    }
+    float temperature = ((float) (i2c_response[0]) * 256 + i2c_response[1]);
```
Comment (jesserockz) [inline esphome/components/hte501/hte501.cpp]:
```suggestion
    float temperature = (float) encode_uint16(i2c_response[0], i2c_response[1]);
```

File: esphome/components/hte501/hte501.cpp [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+    if (temperature > 55536) {
+      temperature = (temperature - 65536) / 100;
+    } else {
+      temperature = temperature / 100;
+    }
+    float humidity = ((float) (i2c_response[3]) * 256 + i2c_response[4]) / 100;
```
Comment (jesserockz) [inline esphome/components/hte501/hte501.cpp]:
```suggestion
    float humidity = ((float) encode_uint16(i2c_response[3], i2c_response[4])) / 100.0f;
```

**Review by jesserockz — CHANGES_REQUESTED** [review]

File: esphome/components/hte501/hte501.cpp (hunk 2 above)
Comment (Stock-M) [reply esphome/components/hte501/hte501.cpp:88]: I changed everything to the crc8 calculation from helpers. When I tested it, it turned out that the wrong checksum was calculated.  I can't find any other crc8 calculation in helpers. Do you have any suggestions.

**Comment by Stock-M** [comment]
I changed everything to the crc8 calculation from helpers. When I tested it, it turned out that the wrong checksum was calculated. I can't find any other crc8 calculation in helpers. Do you have any suggestions.

**Comment by dcousens** [comment]
The code in this repository for `crc8` seems to be an unattributed copy of https://github.com/PaulStoffregen/OneWire/blob/master/OneWire.cpp#L522-L540; and is known as the Dallas / Maxim "OneWire" variant.

This pull request (and https://github.com/esphome/esphome/pull/3773) is additionally using a CRC-8 Dallas / Maxim variant.

The problem is the byte ordering, the code as written by @Stock-M is the same polynomial, but the byte ordering is reversed.
For example, if you changed the `crc8` code (in repository) from

```
  while ((len--) != 0u) {
```

to
```
  for (auto l = 0; l < len; ++l) {
```

Your problem, @Stock-M, would be resolved.  Unfortunately, that change would break others code.

If you're going to fork the in-repository function to add a different byte ordering, you should probably add attribution to the "OneWire" code while you're in there. (Maybe @PaulStoffregen can clarify, I think the CRC part is by Tom Pollard... that library seems to have a decent history!)

**Comment by PaulStoffregen** [comment]
Can't help much.  When I inherited OneWire (about 12-13 years ago) that code was already present and the list of prior authors was already several names.  At the time I tried reaching out to all of them.  Only one replied, to say he was glad someone was going to fix the bugs and hoped it "would not be horrible next time" he used, or something along those lines.  A brief mention of OneWire is probably the proper thing to do, but I wouldn't worry or try to overthink it too much.

**Comment by Stock-M** [comment]
So your suggestion is to add a new crc 8 calculation in helpers with an additional attribute.

**Comment by jesserockz** [comment]
Ok considering the above, lets just leave the crc8 function in this PR and I can look at refactoring it later on.

**Review by jesserockz — APPROVED** [review]

**Pull request merged**